torch/distributed/_checkpoint/resharding.py

```diff
@@ -0,0 +1,197 @@
+        zip(
+            saved_shard.shard_offsets,
+            current_shard.shard_offsets,
+            saved_shard.shard_sizes,
+            current_shard.shard_sizes,
+        )
+    ):
+        min_range_end = min(
+            saved_shard_offset + saved_shard_size,
+            current_shard_offset + current_shard_size,
+        )
+
+        length = min_range_end - max(current_shard_offset, saved_shard_offset)
+
+        if saved_shard_offset > current_shard_offset:
+            offset_for_saved_tensor = 0
+            offset_for_current_tensor = saved_shard_offset - current_shard_offset
+        else:
+            offset_for_saved_tensor = current_shard_offset - saved_shard_offset
+            offset_for_current_tensor = 0
+
+        narrows.append(
+            (dim, offset_for_saved_tensor, offset_for_current_tensor, length)
+        )
+
+    return narrows
+
+
+def _compute_sharded_tensor_md(
+    storage_key_prefix: str, tensor: ShardedTensor
+) -> ExtendedTensorMetadata:
+    smd = []
+    for shard_md in tensor.metadata().shards_metadata:
+        # each shard is in it own storage key.
+        # Most network file system is optimized with single write, multiple read
+        # Unless we can group tensors locally into one big chunk
+        # It might be best to write each shard as one key
+        suffix = "_".join([str(i) for i in shard_md.shard_offsets])
+        storage_key = f"{storage_key_prefix}_{suffix}"
+
+        shard_size = 1
+        for d in shard_md.shard_sizes:
+            shard_size *= d
+
+        # not particularly great
+        storage_size = shard_size * tensor.local_shards()[0].tensor.element_size()
+
+        one_smd = StorageMetadata(
+            shard_metadata=shard_md,
+            storage_key=storage_key,
+            length=storage_size,
+            offset=0,
+        )
+        smd.append(one_smd)
+
+    return ExtendedTensorMetadata(
+        tensor_metadata=tensor.metadata(),
+        storage_metadata=smd,
+    )
+
+
+def prepare_sharded_tensor_write(
+    sharded_tensor: ShardedTensor,
+    storage_key_prefix: str,
+) -> Tuple[List[TensorWriteRequest], ExtendedTensorMetadata]:
+    """
+    Prepare sharded tensor write.
+
+    Args:
+        sharded_tensor: The sharded tensor to persist.
+        storage_key_prefix: The identifier to be prepended to storage keys
+                            associated with the sharded tensor.
+
+    Returns:
+        Write requests for persisting the sharded tensor, and metadata
+        describing the persisted sharded tensor.
+    """
+    write_requests = []
+    for shard in sharded_tensor.local_shards():
+        # each shard has its own storage key.
+        # For most cases, the read is a recovery from a failure to the same sharding
+        # and does not need any resharding, write each shard as is is the most effective
+        suffix = "_".join([str(i) for i in shard.metadata.shard_offsets])
+        storage_key = f"{storage_key_prefix}_{suffix}"
```

Comment:
Changed the separator to`$` and introduced `resharding.STORAGE_KEY_SEPARATOR` constant to track that.
I refactored this into `_get_shard_storage_key` to avoid duplication.